File: Java/백준/Bronze/2231. 분해합/분해합.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    static int selfSum(int x) {
        int s = x;
        while (x > 0) {
            s += x % 10;
            x /= 10;
        }
        return s;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine().trim());
        int ans = 0;

        int d = String.valueOf(N).length();
        int start = Math.max(1, N - 9 * d);
        for (int m = start; m < N; m++) {
            if (selfSum(m) == N) {
                ans = m;
                break;
            }
        }
        System.out.println(ans);
    }
}

File: Python/백준/Bronze/2231. 분해합/분해합.py
import sys
input = sys.stdin.readline

def self_sum(x: int) -> int:
    s = x
    while x > 0:
        s += x % 10
        x //= 10
    return s

def main():
    N = int(input().strip())
    ans = 0
    # 탐색 범위를 줄이기: 최대 자릿수 * 9 만큼 뒤까지만
    strN = str(N)
    d = len(strN)
    start = max(1, N - 9 * d)
    for m in range(start, N):
        if self_sum(m) == N:
            ans = m
            break
    
    print(ans)

if __name__ == "__main__":
    main()